sql: make privilege operations thread-safe

Dedej-Bergin opened the pull request:
Added proper locking to `privilege.go` to fix data race condition.
This required adding the interface `PrivilegeDescriptorInterface`
and the struct `LockedPrivilegeDescriptor` which is similar to the
original `PrivilegeDescriptor` struct except that it has a RWMutex.

```
New Locking Mechanism:

grant -> WriteLock

checkPrivilege -> ReadOnlyLock
```

Fixes: #142992
Epic: CRDB-49398
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146738)
<!-- Reviewable:end -->

Dedej-Bergin committed:
sql: make privilege operations thread-safe

Added proper locking to `privilege.go` to fix data race condition.
This required adding the interface `PrivilegeDescriptorInterface`
and the struct `LockedPrivilegeDescriptor` which is similar to the
original `PrivilegeDescriptor` struct except that it has a RWMutex.

```
New Locking Mechanism:

grant -> WriteLock

checkPrivilege -> ReadOnlyLock
```

Fixes: #142992
Epic: CRDB-49398
Release note: none
